elastic/kibana
# esArchiver datastream support

## Summary

Fixes https://github.com/elastic/kibana/issues/69061

Adds support for archiving/loading/unloading data streams.

When archiving indices we now verify whether we have a backing index of a data stream. When that's the case we build the data stream's backing index template resolving any component templates link, and save it as a `data_stream` record type in the `mappings.json` file:

```
{
  "type": "data_stream",
  "value": {
    "data_stream": "my-data-stream-one",
    "template": {
      "index_patterns": ["my-data-stream-*"],
      "mappings": { ... },
      "settings": { ... },
      ...
    }
  }
}
```
While multiple backing indices of the same data stream can be returned we'll only output a single entry containing the latest mappings and settings.

The documents associated with a data stream keep the same `doc` type but have an additional `data_stream` property used at load time to index it to the appropriate target (we can't directly write to a backing indice) and to pick the correct BULK operation ([data streams can only use create](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html#docs-bulk-api-prereqs)).

**Note:**
The index template could have an ILM policy that we currently don't save or create when loading the archive. We can add it if they are uses case that could benefit from that

### Testing
- the new data streams paths are unit tested
- functional test suites are green
- archived local and cloud data streams

Manual steps
- Create a data stream or target an existing one (see [setup a data stream](https://www.elastic.co/guide/en/elasticsearch/reference/current/set-up-a-data-stream.html))
- Archive it
`node scripts/es_archiver.js save ~/my-data-stream my-data-stream --es-url=http://elastic:changeme@localhost:9200 --kibana-url=http://elastic:changeme@localhost:5601/pat`
- Load the archive (while loading removes existing resources we should test against a clean cluster)
`node scripts/es_archiver.js load ~/my-data-stream --es-url=http://elastic:changeme@localhost:9200 --kibana-url=http://elastic:changeme@localhost:5601/pat`
- Inspect the loaded data stream and template
- Unload the archive
`node scripts/es_archiver.js load ~/my-data-stream --es-url=http://elastic:changeme@localhost:9200 --kibana-url=http://elastic:changeme@localhost:5601/pat`
- Verify the resources are gone
